### ISSUE TITLE
Downloading game through itch client changes filenames with non-ASCII characters

### ISSUE THREAD

**Issue by Omurice2** [itchio/itch#1959]
My game relies on a .txt file named "français" for localization. Playing the game after downloading it from the itch website works fine. Playing the game after downloading it from the client causes a crash because the text file was renamed in the process. The "ç" character changed to "�".

It's really disappointing to see these bugs happening. I shouldn't have to butcher a language to accommodate software.

The client is at version 26.6.3 and I'm using windows 7.

**Comment by fasterthanlime** [comment]
Hey, thanks for the report!

> changes filenames with non-ASCII characters

This depends on a lot of things - it's most probably a .zip file with non-unicode filenames. The client of course has proper unicode support :)

Can you link to your game so I can take a look?

**Comment by Omurice2** [comment]
Sure, it's https://chaikadev.itch.io/type-knight/. Thanks for taking the time to help.

**Comment by fasterthanlime** [comment]
I've taken a look at both zip files on this page, DD21 has:

```
drwxrwxrwx          - Type Knight DD21/polices/
-rwxr-xr-x          - Type Knight DD21/
drwxrwxrwx          - Type Knight DD21/dictionnaires/
-rw-rw-rw- 763.63 KiB Type Knight DD21/audiogroup2.dat
-rw-rw-rw-  14.39 MiB Type Knight DD21/data.win
-rw-rw-rw- 586.40 KiB Type Knight DD21/dictionnaires/en 56k.txt
-rw-rw-rw- 230.79 KiB Type Knight DD21/dictionnaires/fr.txt
-rw-rw-rw-   1.72 KiB Type Knight DD21/Licenses and author links.txt
-rw-rw-rw-  60.36 KiB Type Knight DD21/polices/Moderne Fette Schwabacher.ttf
-rw-rw-rw-  20.41 KiB Type Knight DD21/polices/PixelOperator.ttf
-rw-rw-rw-  19.52 KiB Type Knight DD21/polices/PixelOperator8.ttf
-rw-rw-rw-   4.62 MiB Type Knight DD21/type knight.exe
```

and DD20 has:

```
-rw-rw-rw-  24.64 MiB data.win
-rw-rw-rw-   4.61 MiB type knight.exe
```

I don't see an accent aigu anywhere? Is it in a deleted upload?

**Comment by fasterthanlime** [comment]
Just played a little in kitch v24.40.0 (the itch.io app beta), works well, and is pretty fun too :)

https://i.imgur.com/B2Y7cMm.gif

If you can just confirm that it also works fine for you with https://fasterthanlime.itch.io/install-kitch then we can close this, as it means the issue has been resolved when switching to the new decompression engine (v23 will not be updated as v25 is coming out as soon as the beta ends).

**Comment by Omurice2** [comment]
I renamed the file that was causing the problem yesterday evening. Can you see my hidden files? Try "Type Knight DD21 v1.zip" or "Type Knight DD21 v0.zip". I could make them public again briefly if you need to test things out.

**Comment by Omurice2** [comment]
I renamed "Type Knight DD21 v1.zip" to "Type Knight DD21 v1 (test version).zip" and made it public. I downloaded the game through kitch and the crash still happens.

**Comment by fasterthanlime** [comment]
So, @GranPC and I analyzed the zip:

  * It appears the bad zip has "ç" encoded as [CP-437](https://en.wikipedia.org/wiki/Code_page_437)
  * Extracting the bad zip with windows's built-in zip extractor creates the file with the correct path
  * Zipping the game folder with 7-Zip's GUI creates a bad zip ("ç" encoded as CP-437)

Here's the thing: archiving tools are *notoriously bad* at dealing with non-ASCII encodings.

Here's some of the behaviors we've seen:

  * If all files have ASCII characters, none of the tools have issues
  * If files only have CP-437 characters, apparently 7-zip creates a zip with CP-437 encoding
  * On Japanese Windows, the built-in Windows "Compressed folders" creates a zip with Shift-JIS encoding
  * If files have characters outside of CP-437, 7-zip creates utf-8 zips (proper zips)
  * It's impossible to detect what encoding a file is using, so
    * Shift-JIS zips will only work properly on computers with a Japanese locale, for example
    * Same for CP-437, they'll only work on tools like:
      * Windows's built-in "Compressed folders" feature
      * 7-zip, but *only in GUI mode* - the command-line version will be unable to decode the "ç"

We've experimented in the past with detecting character encoding for zip file entries - not only is this expensive, it's very hit and miss - overall it gave strictly worse results, so the rules for v25 are:

  * zip files must have paths encoded with utf-8, OR have only ASCII characters

I'm not aware of a "good" archiving program, ie. one that always creates zip files with utf-8 encoded paths, except our own `butler mkzip` command (see attached file).

📦 [mkzip.zip](https://github.com/itchio/itch/files/2195075/mkzip.zip)

What I recommend you do - and it pains me that I can't recommend any other option - is to just use https://itch.io/docs/butler/ to upload your game.

I just tested it, and sure enough, it ends up serving a valid zip (with utf-8 paths) that extracts correctly:

  * with the itch.io app
  * with 7-zip (GUI and CLI)
  * with Windows's built-in "Compressed Folders" feature

TL;DR it's 2018 but almost every tool still struggles with encodings - except ours, because they're written in [Go](https://golang.org/), which handles these things correctly. *Please* use butler to upload your game. It's command-line, but the documentation should guide you through it - we're working on a graphical interface to it - here's an [unofficial one](https://seansleblanc.itch.io/gui-butler) if you really really don't want to touch a command prompt.

I'm sorry most archiving tools suck :(

**Comment by Omurice2** [comment]
Wow, I really appreciate such a detailed answer. I was using 7zip and the default windows compression software interchangeably. I wasn't aware they could mess up the encoding. I'm surprised I can't find an option in 7zip to always use utf-8 encoding. Thanks for taking the time to investigate this issue.
